metron-platform/metron-common/src/main/java/org/apache/metron/common/error/MetronError.java

```diff
@@ -0,0 +1,200 @@
+  }
+
+  public MetronError withErrorFields(Set<String> errorFields) {
+    this.errorFields = errorFields;
+    return this;
+  }
+
+
+  public MetronError addRawMessage(Object rawMessage) {
+    if (rawMessage != null) {
+      if (this.rawMessages == null) {
+        this.rawMessages = new ArrayList<>();
+      }
+      this.rawMessages.add(rawMessage);
+    }
+    return this;
+  }
+
+  public MetronError withRawMessages(List<Object> rawMessages) {
+    this.rawMessages = rawMessages;
+    return this;
+  }
+
+  public Optional<Throwable> getThrowable() {
+    return throwable != null ? Optional.of(throwable) : Optional.empty();
+  }
+
+  @SuppressWarnings({"unchecked"})
+  public JSONObject getJSONObject() {
+    JSONObject errorMessage = new JSONObject();
+    errorMessage.put(Constants.SENSOR_TYPE, "error");
+
+		/*
+     * Save full stack trace in object.
+		 */
+    if (throwable != null) {
+      String stackTrace = ExceptionUtils.getStackTrace(throwable);
+      String exception = throwable.toString();
+      errorMessage.put(ErrorFields.EXCEPTION.getName(), exception);
+      errorMessage.put(ErrorFields.STACK.getName(), stackTrace);
+    }
+
+    errorMessage.put(ErrorFields.TIMESTAMP.getName(), System.currentTimeMillis());
+    try {
+      errorMessage.put(ErrorFields.HOSTNAME.getName(), InetAddress.getLocalHost().getHostName());
+    } catch (UnknownHostException ex) {
+
+    }
+    if(rawMessages != null) {
+      for(int i = 0; i < rawMessages.size(); i++) {
+        Object rawMessage = rawMessages.get(i);
+        // If multiple messages are included add an index to the field name, otherwise leave it off
+        String rawMessageField = rawMessages.size() == 1 ? ErrorFields.RAW_MESSAGE.getName() : ErrorFields.RAW_MESSAGE.getName() + "_" + i;
+        String rawMessageBytesField = rawMessages.size() == 1 ? ErrorFields.RAW_MESSAGE_BYTES.getName() : ErrorFields.RAW_MESSAGE_BYTES.getName() + "_" + i;
+        if(rawMessage instanceof byte[]) {
+          errorMessage.put(rawMessageField, Bytes.toString((byte[])rawMessage));
+          errorMessage.put(rawMessageBytesField, toByteArrayList((byte[])rawMessage));
+        } else {
+          errorMessage.put(rawMessageField, rawMessage);
+        }
+      }
+    }
+
+    if (rawMessages != null && rawMessages.size() == 1) {
+      Object rawMessage = rawMessages.get(0);
+      if (rawMessage instanceof JSONObject) {
+        JSONObject rawJSON = (JSONObject) rawMessage;
+        if (errorFields != null) {
+          String errorFieldString = String.join(",", errorFields);
+          errorMessage.put(ErrorFields.ERROR_FIELDS.getName(), errorFieldString);
+          List<String> hashElements = errorFields.stream().map(errorField ->
+                  String.format("%s-%s", errorField, rawJSON.get(errorField))).collect(Collectors.toList());
+          errorMessage.put(ErrorFields.ERROR_HASH.getName(), DigestUtils.sha256Hex(String.join("|", hashElements).getBytes(UTF_8)));
```

Comment:
Hashing was suggested to provide a unique identifier of the original information, and thus giving you something to key in on and create a count of unique events/prioritize more prevalent issues without the concern of cyclical issues (for instance, if the issue is with indexing a specific message, and you try to index it again, it will just fail in a loop).